packages/program-boilerplate/src/conversion.ts

```diff
@@ -66,6 +69,47 @@ function parseValue(value: string): RuleValue {
   return value;
 }
 
+/**
+ * Checks whether any of the edge trigger conditions are met given
+ * a list of fields to check and the current context.
+ *
+ * @param {string[] | undefined} fields The list of edge trigger fields
+ * @param {any} activeTrigger The current program `activeTrigger`
+ *
+ * @return {boolean} Whether any of the edge fields have changed
+ */
+export function meetEdgeTriggerConditions(
+  fields: string[] | undefined,
+  activeTrigger: any
+): boolean {
+  if (fields === undefined || fields.length === 0) {
+    return true;
+  }
+
+  for (const field of fields) {
+    if (!field.startsWith("user.")) {
+      // TODO: what to do here? this is probably some kind of error case
+      continue;
+    }
+
+    const previousValue = jsonata(field.replace("user.", "previous.")).evaluate(
+      activeTrigger
+    );
+    const currentValue = jsonata(field).evaluate(activeTrigger);
```

Comment:
@jayden-chan These calls to jsonata could fail, you might want to wrap them in try/catch.

Also, while it's unlikely, there's a possibility to cause a denial of service by passing a badly behaved expression. For example, something like `user.blah or ($f := function() { $f() }; $f())` would actually cause an infinite loop.

For this reason please use `safeJsonata`.

Reply:
@johanventer I refactored this to use the safe JSONata, but I had to refactor the function signature and it's not backward compatible with the old `safeJsonata`. Previously it was not possible to determine the difference between a successful JSONata that returned `undefined` and a failed JSONata. I need to know this difference for the edge trigger stuff so I just made a new function temporarily. I guess we need to decide if I should just leave the second function in there or if we should increment the major version of program boilerplate.

Reply:
Thanks @jayden-chan I missed this comment, so my other comment about `safeJsonata2` was based on false assumptions.